subprojects/snapshots/src/main/java/org/gradle/internal/vfs/impl/DefaultWatchingVirtualFileSystem.java

```diff
@@ -80,55 +116,86 @@ public void handleLostState() {
         }
     }
 
-    @Override
-    public void stopWatching() {
+    /**
+     * Stop watching the known areas of the file system, and invalidate
+     * the parts that have been changed since calling {@link #startWatching(Collection)} ()}.
+     */
+    private void stopWatching() {
         if (watchRegistry == null) {
             return;
         }
 
         try {
-            long startTime = System.currentTimeMillis();
-            // Make sure events stop arriving before obtaining the statistics
             watchRegistry.close();
-            FileWatcherRegistry.FileWatchingStatistics statistics = watchRegistry.getStatistics();
+        } catch (IOException ex) {
+            LOGGER.error("Couldn't fetch file changes, dropping VFS state", ex);
+            invalidateAll();
+        } finally {
+            watchRegistry = null;
+        }
+    }
+
+    private void handleWatcherRegistryEvents(String eventsFor) {
+        if (watchRegistry != null) {
+            FileWatcherRegistry.FileWatchingStatistics statistics = watchRegistry.getAndResetStatistics();
+            LOGGER.warn("Received {} file system events {}", statistics.getNumberOfReceivedEvents(), eventsFor);
             if (statistics.isUnknownEventEncountered()) {
                 LOGGER.warn("Dropped VFS state due to lost state");
             }
             if (statistics.getErrorWhileReceivingFileChanges().isPresent()) {
                 LOGGER.warn("Dropped VFS state due to error while receiving file changes", statistics.getErrorWhileReceivingFileChanges().get());
+                invalidateAll();
             }
-            LOGGER.warn("Received {} file system events since last build", statistics.getNumberOfReceivedEvents());
-            LOGGER.warn("Spent {} ms processing file system events since last build", System.currentTimeMillis() - startTime);
-        } catch (IOException ex) {
-            LOGGER.error("Couldn't fetch file changes, dropping VFS state", ex);
-            invalidateAll();
-        } finally {
-            watchRegistry = null;
         }
     }
 
+    private void printStatistics(String verb, String statisticsFor) {
+        VirtualFileSystemStatistics statistics = getStatistics();
+        LOGGER.warn(
+            "Virtual file system {} information about {} files, {} directories and {} missing files {}",
+            verb,
+            statistics.getRetained(FileType.RegularFile),
+            statistics.getRetained(FileType.Directory),
+            statistics.getRetained(FileType.Missing),
+            statisticsFor
+        );
+    }
+
     @Override
-    public VirtualFileSystemStatistics getStatistics() {
+    public void update(Iterable<String> locations, Runnable action) {
+        if (buildRunning) {
+            producedByCurrentBuild.updateAndGet(currentValue -> {
```

Comment:
Why do we need to distinguish if a thing was produced during the current build? Just because the event was received between the start and end of the build doesn't mean it was _produced_ by it. Can we remove this state somehow?

Reply:
If a part of the VFS is invalidated by a call to `update` from Gradle code, then the current build modifies that part of the file system. So I think `producedByCurrentBuild` makes sense. `modifiedByCurrentBuild` would be an alternative.

Note that no events from watchers cause an update to `producedByCurrentBuild`. All those updates directly modify the state of the VFS.

Instead of trying to filter the events, we could try to remove unregister the watchers. Would that be better?

Reply:
Ah, we can't unregister the watchers on Windows and macOS. That is why I started filtering the events.